compiler/src/vm/heap/object.rs

```diff
@@ -0,0 +1,257 @@
+use super::{pointer::Pointer, Heap};
+use crate::{
+    builtin_functions::BuiltinFunction,
+    compiler::{
+        hir_to_lir::HirToLir,
+        lir::{Instruction, Lir},
+    },
+    database::Database,
+    module::Module,
+};
+use itertools::Itertools;
+use num_bigint::BigInt;
+use std::{
+    collections::{hash_map::DefaultHasher, HashMap},
+    hash::{Hash, Hasher},
+    ops::Deref,
+};
+
+#[derive(Clone)]
+pub struct Object {
+    pub reference_count: usize,
+    pub data: Data,
+}
+#[derive(Clone)]
+pub enum Data {
+    Int(Int),
+    Text(Text),
+    Symbol(Symbol),
+    Struct(Struct),
+    Closure(Closure),
+    Builtin(Builtin),
+}
+
+#[derive(Clone)]
+pub struct Int {
+    pub value: BigInt,
+}
+
+#[derive(Clone)]
+pub struct Text {
+    pub value: String,
+}
+
+#[derive(Clone)]
+pub struct Symbol {
+    // TODO: For better memory efficiency, store the symbol as a null-terminated
+    // string.
```

Comment:
Because it's a subset of ASCII, we could store it even more efficiently (5 bits for the first and 6 bits for the following characters are sufficient), and we should probably intern them